### PR TITLE
Implement Public API and Tap on top of Lister

### PR DESCRIPTION
**Pull request by siggy**
public-api and and tap were both using their own implementations of
the Kubernetes Informer/Lister APIs.

This change factors out all Informer/Lister usage into the Lister
module. This also introduces a new `Lister.GetObjects` method.

Signed-off-by: Andrew Seigner <siggy@buoyant.io>

### CODE REVIEW

File: cli/Dockerfile-bin [hunk 1]
```diff
@@ -1,8 +1,9 @@
 ## compile binaries
-FROM gcr.io/runconduit/go-deps:a6e8221d as golang
+FROM gcr.io/runconduit/go-deps:ff0dc047 as golang
 ARG CONDUIT_VERSION
 WORKDIR /go/src/github.com/runconduit/conduit
 COPY cli cli
+COPY controller/k8s controller/k8s
```
Comment (olix0r) [inline cli/Dockerfile-bin]: is this for the parsing utilities? does the controller actually need the parsing stuff? can they be moved to the cli package?  I'm looking to minimize docker cache invalidation for the CLI...

File: Gopkg.lock [hunk 2]
```diff
@@ -746,6 +746,6 @@
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
-  inputs-digest = "325d0175a74db86d1fed482bddc8279459d63be9d5338d9a420722c24eea5534"
+  inputs-digest = "182946acde1cd53f9b8e6e3137ea55d1d14d7467a175ce261ddf30d74de3168e"
```
Comment (olix0r) [inline Gopkg.lock]: i still don't quite understand why this changes every time even though we havent' changed gopkg.toml

Comment (siggy) [reply Gopkg.lock]: running `bin/dep hash-inputs` on master and this branch reveals two additional dependencies:
k8s.io/apimachinery/pkg/api/meta
k8s.io/apimachinery/pkg/runtime/schema

while both of these dependencies are already referenced via:
```
[[constraint]]
  name = "k8s.io/apimachinery"
  version = "kubernetes-1.9.1"
```
... we are still creating a new dependency in our code on those two packages.

File: cli/Dockerfile-bin (hunk 1 above)
Comment (siggy) [reply cli/Dockerfile-bin]: Unfortunately it's `controller/api/public/grpc_server.go` depending on `controller/k8s.Lister`. Since `Dockerfile-bin` already pulls in `controller/api`, it now must pull in `controller/k8s`:
https://github.com/runconduit/conduit/blob/38cc23e67e521eff0e491cb04263129040fb3d1d/controller/api/public/grpc_server.go#L14
